File: 367.valid-perfect-square.cpp
/*
 * @lc app=leetcode id=367 lang=cpp
 *
 * [367] Valid Perfect Square
 */

#include <bits/stdc++.h>

#include <cassert>

using namespace std;

// @lc code=start
class Solution {
 public:
  bool isPerfectSquare(int num) { return (int)pow((int)sqrt(num), 2) == num; }
};
// @lc code=end

int main() {
  Solution s;
  assert(s.isPerfectSquare(0));
  assert(s.isPerfectSquare(15) == false);
  assert(s.isPerfectSquare(16));
  assert(s.isPerfectSquare(17) == false);
  assert(s.isPerfectSquare(2) == false);
  assert(s.isPerfectSquare(1));
  return 0;
}

File: 367.valid-perfect-square.rs
/*
 * @lc app=leetcode id=367 lang=rust
 *
 * [367] Valid Perfect Square
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn is_perfect_square(num: i32) -> bool {
        let num = num as i64;
        for i in 1..num {
            let sq: i64 = i * i;
            if sq < num {
                continue;
            } else {
                return sq == num;
            }
        }
        true
    }
}
// @lc code=end

#[test]
fn test1() {
    assert!(Solution::is_perfect_square(16));
}

#[test]
fn test2() {
    assert!(!Solution::is_perfect_square(14));
}
